File: LeetCode/src/main/java/com/jack/bit/LeetCode136.py

from typing import List


class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        a_set = set()
        res = 0
        
        for num in nums:
            if num in a_set:
                res-=num
            else :
                res+=num
                a_set.add(num)
        return res
        

class BetterSolution:
    def singleNumber(self, nums: List[int]) -> int:
        res = 0
        for num in nums:
            res ^= num
            
        return res

if __name__ == "__main__":
        
        nums = [1,1,2,3,3,4,4]
        res = BetterSolution().singleNumber(nums)
        print(res)
    
    
    

File: LeetCode/src/main/java/com/jack/bit/LeetCode136.java
package com.jack.bit;

import java.util.HashSet;
import java.util.Set;

public class LeetCode136 {

    public static void main(String[] args) {

        BetterSolution solution = new BetterSolution();
        // Solution solution = new Solution();
        int [] nums = {1,1,2,2,3,3,4};
        int res =solution.singleNumber(nums);

        System.out.println(res);
        
    }


    static class BetterSolution{
        public int singleNumber(int[] nums) {
            int res =0;
            for(int num: nums){
                res ^= num;
            }
            return res;
       
        }

    }


    static class Solution{
        public int singleNumber(int[] nums) {
            int res =0;
            Set<Integer> set = new HashSet<>();

            for (int num : nums) {
                if(set.contains(num)){
                    res -= num;
                }else{
                    res += num;
                    set.add(num);
                }
            }
            return res;
        }
    }


}
